src/mpi/romio/adio/common/ad_resize.c

```diff
@@ -27,6 +27,7 @@ void ADIOI_GEN_Resize(ADIO_File fd, ADIO_Offset size, int *error_code)
 
     /* --BEGIN ERROR HANDLING-- */
     if (err == -1) {
+        MPI_Bcast(&errno, 1, MPI_INT, fd->hints->ranklist[0], fd->comm);
```

Comment:
Yes, the bcast just before the ``if (err == -1)`` ensures that ``err = -1`` on all ranks, so the second broadcast will not hang.  However, the MPI error code is then computed using the value of ``errno`` on this line:
```
 *error_code = ADIOI_Err_create_code(myname, fd->filename, errno);
```
Without the second bcast, ``errno`` is only well-defined on the single rank that called ftruncate.  This second bcast is needed to ensure all ranks have ``errno`` set correctly.

Reply:
I think you can skip the `Bcast` and call ` *error_code = ADIOI_Err_create_code(myname, fd->filename, errno);` directly. `error_code` is for local process error checking only. There isn't requirement that all process have to have the same error code.

Reply:
Yes, I suppose that does fit with my understanding of error codes returned by other MPI collectives.  In that case, isn't the existing bcast of ``err`` unnecessary?

Reply:
After double checking the code and your issue comments, the problem here is we propagated the `err` but not the `errno` (reason of error). Is it ok for non-zero rank to simply use a error no such as `EOTHER` (you'll need to add it in `adio/common/error.c`.

My concern is when there is error, if you don't know the exact the reason such as the case of non-zero ranks, running another collective such as `bcast` may hang due to this unknown error.

Reply:
If when `err == -1`, `errno` is always non-zero -- seems to be the case, the first `Bcast` should simply bcast `errno` and the just check `errno` in the ERROR HANDLING block.

Reply:
Yes, for that approach, I'd recommend going with something like the alternate implementation I show in the comments near the top.  It's the same idea as what you're suggesting, but I copy errno into err first.

It also now occurs to me that broadcasting the errno variable directly could be risky, since it may be possible that some other libc call within the MPI stack overwrites errno before its value is actually sent out on the wire.

Reply:
> Yes, for that approach, I'd recommend going with something like the alternate implementation I show in the comments near the top. It's the same idea as what you're suggesting, but I copy errno into err first.

Yes, (sorry I was slow to catch up), do that.

Reply:
Ok, I just force pushed back with this other version.